halo/evmstaking/evmstaking.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/omni-network/omni/lib/ethclient"
+	"github.com/omni-network/omni/lib/k1util"
+	"github.com/omni-network/omni/lib/log"
+
+	k1 "github.com/cometbft/cometbft/crypto/secp256k1"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	akeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	bkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	skeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+const AccountName = "evmstaking"
+
+var _ evmenginetypes.EvmLogProvider = LogProvider{}
+
+// LogProvider implements the evmenginetypes.EvmLogProvider interface.
+type LogProvider struct {
+	contract *bindings.OmniStake
+	ethCl    ethclient.Client
+	address  common.Address
+	sKeeper  *skeeper.Keeper
+	bKeeper  bkeeper.Keeper
+	aKeeper  akeeper.AccountKeeper
+}
+
+// New returns a new LogProvider.
+func New(
+	ethCl ethclient.Client,
+	sKeeper *skeeper.Keeper,
+	bKeeper bkeeper.Keeper,
+	aKeeper akeeper.AccountKeeper,
+) (LogProvider, error) {
+	address := common.HexToAddress(predeploys.OmniStake)
+	contract, err := bindings.NewOmniStake(address, ethCl)
+	if err != nil {
+		return LogProvider{}, errors.Wrap(err, "new omni stake")
+	}
+
+	return LogProvider{
+		contract: contract,
+		ethCl:    ethCl,
+		address:  address,
+		sKeeper:  sKeeper,
+		bKeeper:  bKeeper,
+		aKeeper:  aKeeper,
+	}, nil
+}
+
+// Logs returns all omni stake contract EVM logs from the provided block hash.
+func (l LogProvider) Logs(ctx context.Context, blockHash common.Hash) ([]*evmenginetypes.EVMLog, error) {
+	// TODO(corver): Maybe we should filter by expected topic only.
+	logs, err := l.ethCl.FilterLogs(ctx, ethereum.FilterQuery{
+		BlockHash: &blockHash,
+		Addresses: l.Addresses(),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "filter logs")
+	}
+
+	resp := make([]*evmenginetypes.EVMLog, 0, len(logs))
+	for _, l := range logs {
+		topics := make([][]byte, 0, len(l.Topics))
+		for _, t := range l.Topics {
+			topics = append(topics, t.Bytes())
+		}
+		resp = append(resp, &evmenginetypes.EVMLog{
+			Address: l.Address.Bytes(),
+			Topics:  topics,
+			Data:    l.Data,
+		})
+	}
+
+	return resp, nil
+}
+
+func (l LogProvider) Addresses() []common.Address {
+	return []common.Address{l.address}
+}
+
+// DeliverLog processes a omni deposit log event:
+// - Mint the corresponding amount of $STAKE coins.
+// - Send the minted coins to the depositor's account.
+// - Create a new validator with the depositor's account.
+func (l LogProvider) DeliverLog(ctx context.Context, _ common.Hash, elog *evmenginetypes.EVMLog) error {
+	deposit, err := l.contract.ParseDeposit(elog.ToEthLog())
+	if err != nil {
+		return errors.Wrap(err, "parse deposit")
+	}
+
+	log.Info(ctx, "EVM staking deposit detected, adding new validator",
+		"depositor", deposit.Depositor.Hex(),
+		"amount", deposit.Amount.String())
```

Comment:
Flagging - this could be an existing validator adding stake

Reply:
Also - remove `Depositor` and replaced it with `Pubkey`.  We can add it back. Tho cchain should not need it. Could be useful for offchain tools to index val deposits by events.

Reply:
yup, still need to add logic for "adding more stake" as opposed to "adding initial stake".